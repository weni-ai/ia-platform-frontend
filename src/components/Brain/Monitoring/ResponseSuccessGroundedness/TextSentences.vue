<template>
  <section
    data-testid="sentences"
    class="text-sentences"
    @mouseleave="clearHoveredSentence"
  >
    <UnnnicToolTip
      v-for="(fragment, index) in fragments"
      :key="fragment.sentence"
      data-testid="sentence-text"
      :text="getTooltipText(fragment)"
      :class="getTextClasses(index, fragment)"
      side="top"
      enabled
      :forceOpen="hoveredSentence === index"
      @mouseover="setHoveredSentence(index)"
    >
      {{ fragment.sentence }}
    </UnnnicToolTip>
  </section>
</template>

<script setup>
import { ref } from 'vue';

import i18n from '@/utils/plugins/i18n';

const props = defineProps({
  fragments: {
    type: Array,
    required: true,
  },
  getReliabilityLevel: {
    type: Function,
    required: true,
  },
});

const hoveredSentence = ref(null);

function setHoveredSentence(index) {
  hoveredSentence.value = index;
}

function clearHoveredSentence() {
  hoveredSentence.value = null;
}

function truncateString(string, maxLength = 25) {
  if (!string) return;

  return string.length > maxLength
    ? `${string.slice(0, maxLength)}...`
    : string;
}

function getTooltipText(fragment) {
  const truncatedFileName = truncateString(fragment.sources?.[0].filename);
  const confidence = i18n.global.t(
    'router.monitoring.inspect_response.confidence_percentage',
    {
      percentage: fragment.score,
    },
  );
  return `${truncatedFileName}\n(${confidence})`;
}

function getTextClasses(index, fragment) {
  return [
    'text-sentences__text',
    `text-sentences__text--${props.getReliabilityLevel(fragment)}`,
    { 'text-sentences__text--focused': hoveredSentence.value === index },
  ];
}
</script>

<style scoped lang="scss">
.text-sentences {
  .text-sentences__text {
    margin-right: $unnnic-spacing-nano;

    display: inline;

    line-height: $unnnic-font-size-body-lg + $unnnic-line-height-md;
    color: $unnnic-color-neutral-darkest;
    font-size: $unnnic-font-size-body-lg;
    text-decoration: underline;
    text-decoration-color: $unnnic-color-neutral-cleanest;
    text-decoration-thickness: 3px;
    text-underline-offset: 4px;

    cursor: default;

    &--focused {
      text-decoration-thickness: 5px;
    }

    &--slightly-reliable {
      text-decoration-color: $unnnic-color-weni-300;
    }
    &--moderately-reliable {
      text-decoration-color: $unnnic-color-weni-600;
    }
    &--highly-reliable {
      text-decoration-color: $unnnic-color-weni-800;
    }
  }
}
</style>
