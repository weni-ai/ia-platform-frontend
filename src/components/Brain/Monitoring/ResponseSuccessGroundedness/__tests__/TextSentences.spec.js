import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { mount } from '@vue/test-utils';

import TextSentences from '../TextSentences.vue';

describe('TextSentences.vue', () => {
  let wrapper;

  const fragmentsMock = [
    {
      sentence: 'This is slightly reliable.',
      sources: [{ filename: 'Test filename' }],
      score: 0,
      reliability: 'slightly-reliable',
    },
    {
      sentence: 'This is moderately reliable.',
      sources: [{ filename: 'Test filename' }],
      score: 0,
      reliability: 'moderately-reliable',
    },
    {
      sentence: 'This is highly reliable.',
      sources: [{ filename: 'Test filename' }],
      score: 0,
      reliability: 'highly-reliable',
    },
  ];

  const getReliabilityLevelMock = vi.fn((fragment) => fragment.reliability);

  const mountWrapper = (props = {}) => {
    wrapper = mount(TextSentences, {
      props: {
        fragments: fragmentsMock,
        getReliabilityLevel: getReliabilityLevelMock,
        ...props,
      },
    });
  };

  beforeEach(() => {
    mountWrapper();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  const sentences = () => wrapper.findAll('[data-testid="sentence-text"]');

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders a section with the correct class', () => {
    const sentencesContainer = wrapper.find('[data-testid="sentences"]');
    expect(sentencesContainer.exists()).toBe(true);
  });

  it('renders the correct number of sentences for each fragment', () => {
    expect(sentences().length).toBe(fragmentsMock.length);
  });

  it('renders each sentence correctly', () => {
    sentences().forEach((sentence, index) => {
      expect(sentence.text()).toContain(fragmentsMock[index].sentence);
    });
  });

  it('renders each sentence tooltip correctly', () => {
    sentences().forEach((sentence, index) => {
      expect(sentence.text()).toContain(
        wrapper.vm.getTooltipText(fragmentsMock[index]).replace('\n', ' '),
      );
    });
  });

  it('sets hoveredSentence on mouseover and clears it on mouseleave', async () => {
    await sentences()[0].trigger('mouseover');
    expect(wrapper.vm.hoveredSentence).toBe(0);

    await sentences()[1].trigger('mouseover');
    expect(wrapper.vm.hoveredSentence).toBe(1);

    await wrapper.find('[data-testid="sentences"]').trigger('mouseleave');
    expect(wrapper.vm.hoveredSentence).toBeNull();
  });

  it('applies the correct dynamic class for each fragment', () => {
    sentences().forEach((sentence, index) => {
      const reliabilityClass = `text-sentences__text--${fragmentsMock[index].reliability}`;
      expect(sentence.classes()).toContain(reliabilityClass);
    });
  });

  it('calls getReliabilityLevel for each fragment', () => {
    expect(getReliabilityLevelMock).toHaveBeenCalledTimes(fragmentsMock.length);
    fragmentsMock.forEach((fragment) => {
      expect(getReliabilityLevelMock).toHaveBeenCalledWith(fragment);
    });
  });

  it('renders no sentence when fragments are empty', async () => {
    await wrapper.setProps({ fragments: [] });
    expect(sentences().length).toBe(0);
  });

  it('renders different reliability levels correctly', () => {
    const reliabilityClasses = [
      'text-sentences__text--slightly-reliable',
      'text-sentences__text--moderately-reliable',
      'text-sentences__text--highly-reliable',
    ];
    sentences().forEach((sentence, index) => {
      expect(sentence.classes()).toContain(reliabilityClasses[index]);
    });
  });

  describe('truncateString', () => {
    const truncateString = (str, size) => wrapper.vm.truncateString(str, size);

    it('should return undefined for empty string', () => {
      expect(truncateString('')).toBeUndefined();
    });

    it('should return the original string if it is shorter than the maxLength', () => {
      expect(truncateString('Hello, world!')).toBe('Hello, world!');
    });

    it('should truncate the string if it is longer than the maxLength', () => {
      expect(truncateString('This is a very very very long string', 10)).toBe(
        'This is a ...',
      );
    });

    it('should use the default maxLength if not provided', () => {
      expect(truncateString('This is a very very very long string')).toBe(
        'This is a very very very ...',
      );
    });

    it('should handle strings with exact maxLength', () => {
      expect(truncateString('123456789012345', 15)).toBe('123456789012345');
    });
  });
});
