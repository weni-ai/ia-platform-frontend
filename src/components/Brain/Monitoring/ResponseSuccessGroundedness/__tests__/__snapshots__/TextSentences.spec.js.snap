// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextSentences.vue > matches the snapshot 1`] = `
"<section data-v-44bdc72e="" data-testid="sentences" class="text-sentences">
  <div data-v-5b8c172b="" data-v-44bdc72e="" class="unnnic-tooltip text-sentences__text text-sentences__text--slightly-reliable" data-testid="sentence-text">This is slightly reliable.<span data-v-5b8c172b="" class="unnnic-tooltip-label unnnic-tooltip-label-top" data-testid="tooltip-label" style="left: 0px; top: -8px;">Test filename <br data-v-5b8c172b="">(0% confidence) <br data-v-5b8c172b=""><!----></span></div>
  <div data-v-5b8c172b="" data-v-44bdc72e="" class="unnnic-tooltip text-sentences__text text-sentences__text--moderately-reliable" data-testid="sentence-text">This is moderately reliable.<span data-v-5b8c172b="" class="unnnic-tooltip-label unnnic-tooltip-label-top" data-testid="tooltip-label" style="left: 0px; top: -8px;">Test filename <br data-v-5b8c172b="">(0% confidence) <br data-v-5b8c172b=""><!----></span></div>
  <div data-v-5b8c172b="" data-v-44bdc72e="" class="unnnic-tooltip text-sentences__text text-sentences__text--highly-reliable" data-testid="sentence-text">This is highly reliable.<span data-v-5b8c172b="" class="unnnic-tooltip-label unnnic-tooltip-label-top" data-testid="tooltip-label" style="left: 0px; top: -8px;">Test filename <br data-v-5b8c172b="">(0% confidence) <br data-v-5b8c172b=""><!----></span></div>
</section>"
`;
